File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int buyDay = prices[0], profit = 0;
        
        for (int day = 1; day < prices.length; day++) {
            
            int curProfit = prices[day] - buyDay;
            
            profit = Math.max(profit, curProfit);
            
            buyDay = Math.min(buyDay, prices[day]);
        }
        
        return profit;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int profit = 0, buy = 0;
        
        for (int sell = 1, n = prices.size(); sell < n; sell++) {
            profit = max(profit, prices[sell] - prices[buy]);
            
            if (prices[sell] < prices[buy]) {
                buy = sell;
            }
        }
        
        return profit;
    }
};